tensorflow/lite/micro/docs/rfc-16x8-quantization-port.md

```diff
@@ -0,0 +1,138 @@
+
+> [16x8 quantization] mode can improve accuracy of the quantized model significantly, when activations are sensitive to the quantization, while still achieving almost 3-4x reduction in model size. Moreover, this fully quantized model can be consumed by integer-only hardware accelerators.
+
+Edge Impulse, a company that deploys TensorFlow Lite for Microcontrollers as part of its embedded
+machine learning pipeline, has gathered feedback from customers with production models for which 8 bit
+quantization results in unacceptable degradation but for whom 16x8 is fine.
+
+While 16x8 quantization is well supported within TensorFlow Lite, it is not currently supported
+within TensorFlow Lite for Microcontrollers. Porting the TensorFlow Lite reference kernels is
+relatively straightforward and will improve adoption of TensorFlow Lite for Microcontrollers with users
+for whom degradation is too severe with full 8 bit quantization.
+
+## User Benefit
+
+The headline would be "16x8 kernels improve accuracy for quantized models on microcontrollers without
+increasing model size".
+
+Users would benefit in the following ways:
+
+- Improved accuracy for quantized models without increasing model size (in exchange for additional
+  runtime memory usage)
+- Improved performance under certain conditions (for example, 16x8 CMSIS-NN kernels will run faster)
+  than 8 bit kernels since less unpacking is required)
+
+## Design Proposal
+
+This is the meat of the document, where you explain your proposal. If you have
+multiple alternatives, be sure to use sub-sections for better separation of the
+idea, and list pros/cons to each approach. If there are alternatives that you
+have eliminated, you should also list those here, and explain why you believe
+your chosen approach is superior.
+
+Make sure you’ve thought through and addressed the following sections. If a section is not relevant to your specific proposal, please explain why, e.g. your RFC addresses a convention or process, not an API.
+
+
+We propose that the 16x8 kernels are ported from the TensorFlow Lite reference kernels to
+TensorFlow Lite for Microcontrollers following the process in the [Porting TensorFlow Lite Ops to Micro](https://docs.google.com/document/d/1KLJTPWm4TUKB9YyIqFJl9VCP0ZMJDt_P8RNpRmwqMxw/edit#heading=h.5x0d5h95i329)
+guide.
+
+We wish to ensure that the following kernels are compatible with 16x8 mode:
+
+- Conv2D
+- MaxPool2D
+- DepthwiseConv2D
+- FullyConnected
+- Relu
+- Relu6
+- Tanh
+- Softmax
+- Pad
+- Reshape
+- Pack
+- Unpack
+- Add
+- Mul
+
+Adding the 16x8 kernels directly to TFLM alongside the existing kernels would increase the default code size by an unacceptable amount. Instead, we will make use of the kernel registration API currently under development by the TFLM team. The use of this is demonstrated in the
+[Keyword benchmark code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/benchmarks/keyword_benchmark.cc#L56).
```

Comment:
use a permalink instead:

https://github.com/tensorflow/tensorflow/blob/a30d20b632b4ffbfd437ccf8ee205fef0917a3eb/tensorflow/lite/micro/benchmarks/keyword_benchmark.cc#L56

Reply:
In this current this example, there are two ways to register the operator; "INT8 only" or "all supported data types". How would the "all supported data types" be handled for the INT16 work covered by this RFC. Would it include INT8, INT16 and FLOAT? If yes, this would increase the code size for users who aren't using the explicit "Register_OPERATOR_INT8" call. 

We do see usecases where an "all supported data types" could be useful. For example, if TFLM is running on a small microcontroller-like subsystem that need to be capable of handling any type of network thown at it - typically by a larger subsystem...

This also raises the question if it should be possible to register multiple variants of an operator. Like this:
```
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8()); 
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT16());
``` 
I may be wrong, but this might be tricky to solve with the current architecture...

Thoughts?

Reply:
You are correct, for the case of registering all supported data types, the code size will increase. We are implicitly saying that that is acceptable from the perspective of the TFLM platform. If your application wants to support any type of network, then you will see a binary size increase as the types of networks that TFLM supports increases.

What we do want to enable is for applications to avoid this additional cost (explicitly).

And you are also correct that
```
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8()); 
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT16());
```
isn't really possible with the current way of writing kernels or the surrounding APIs/architecture.


With the variant registration API, you could add something like (somewhat awkward, and would need proper justification, but feasible):
```
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8_INT16());
```

Our point of view on this topic is that if your application code and model are intertwined (i.e. you know exactly what model you'd like to run, update them together ...), then TFLM would like to enable you to only pay for what you need (we haven't met that goal today, but would like to move towards it instead of away from it).

For a more general-purpose application, you get all the bells and whistles and associate code size increase.

Reply:
Sounds good to me. To summarize, the application has a few options:

1 ) Register the all ops resolver. Every op and every supported datatype of the op is part of the final binary. 

2a ) Register the mutable ops resolver and register the all datatypes version. Example:
`op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED);`

2b ) Register the mutable ops resolver and register a single datatype version. Example:
`op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8);`

2a and 2b would be done per operator basis.

* Option 1 would be the most expensive in terms of memory, but most flexible in terms of support.
* Option 2a can be used as a middle ground between 1 and 2b. If you, for example, have a model with multiple datatypes (possible?) or if your app has multiple models containing multiple datatypes. 
* Option 2b would be the most optimal solution in terms of memory, as you pay only for what you need.

Does this make sense?

I agree that
`op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8_INT16());`
would need proper justification since it adds a penalty to the readability and understandability of the code.

Reply:
Thanks for the concise summary!

Reply:
In the PR I suggested the following suffix for 16x8 quantization, since it seems more clear than `INT8_INT16`:

```
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT16X8());
```

What do you think?

Reply:
sgtm

Reply:
@dansitu thx for the update! The `_INT8_INT16` suffix mentioned above was an example for when both int8 (8x8) and int16 (16x8) quantized kernels are registered. Related to this comment:

> This also raises the question if it should be possible to register multiple variants of an operator. Like this:
> `op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8());`
> `op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT16());`

However, such a usecase would need justification.

I don't have a strong opinion about using `_INT16X8` or `_INT16` suffix for 16x8 quantization. Either works for me.

Reply:
Got it, thank you for the clarification. With that understood I don't have a strong preference on `INT16` vs `INT16x8`.

It seems like being able to register multiple variants would be helpful in keeping the code size down. What's your opinion @advaitjain?

Reply:
let's go with `INT16X8` to stay consistent with the terminology used in the TfLite quantization spec.

Reply:
Sounds good. From the above, it seems like we are comfortable with making it impossible to load multiple variants of the same kernel (e.g. INT8 and INT16 simultaneously)?

Reply:
not impossible, but certainly very deliberate and not out-of-the-box. The variant registration feature is very new and not really used in production yet so we're going to take it slowly.

Per Fredrik's [summarization](https://github.com/tensorflow/tensorflow/pull/46767#discussion_r577427812) in this thread, we can consider

```op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED_INT8_INT16X8());```

but answering whether that is worth the complexity or not is something we will revisit later.

Reply:
@advaitjain @dansitu 

> Sounds good. From the above, it seems like we are comfortable with making it impossible to load multiple variants of the same kernel (e.g. INT8 and INT16 simultaneously)?

I'd say 

 i) pick one version of the kernel (e.g. "Register_FULLY_CONNECTED_INT8") or 
 ii) pick all versions (e.g. "Register_FULLY_CONNECTED"). Includes datatypes float, int8, int16 quantizations.

would be a highly useful usecase. The latter, ii), was what I meant with this option:

> 2a ) Register the mutable ops resolver and register the all datatypes version. Example:
op_resolver->AddFullyConnected(tflite::Register_FULLY_CONNECTED);

Let me know what you think.